File: android/app/src/main/java/com/konrad/nospam/ReportService.kt
package com.konrad.nospam

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

enum class ReportCategory(val apiValue: String) {
    SPAM("spam"),
    SCAM("scam"),
    TELEMARKETING("telemarketing"),
    ROBOCALL("robocall"),
    OTHER("other"),
}

class ReportSubmissionException(val statusCode: Int) : Exception("Supabase returned HTTP $statusCode")

object ReportService {
    suspend fun submit(context: Context, phoneNumberE164: String, category: ReportCategory) {
        withContext(Dispatchers.IO) {
            val connection = URL("${SupabaseConfig.URL}/rest/v1/reports").openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                connection.setRequestProperty("apikey", SupabaseConfig.PUBLISHABLE_KEY)
                connection.setRequestProperty("Authorization", "Bearer ${SupabaseConfig.PUBLISHABLE_KEY}")
                connection.setRequestProperty("Content-Type", "application/json")
                connection.setRequestProperty("Prefer", "return=minimal")
                connection.doOutput = true

                val payload = JSONObject().apply {
                    put("phone_number", phoneNumberE164)
                    put("category", category.apiValue)
                    put("device_hash", DeviceIdentity.hashedDeviceId(context))
                    put("app_platform", "android")
                }

                OutputStreamWriter(connection.outputStream, Charsets.UTF_8).use { it.write(payload.toString()) }

                val statusCode = connection.responseCode
                if (statusCode !in 200..299) {
                    throw ReportSubmissionException(statusCode)
                }
            } finally {
                connection.disconnect()
            }
        }
    }
}

File: ios/NoSpam/ReportService.swift
import Foundation

enum ReportCategory: String, CaseIterable {
    case spam, scam, telemarketing, robocall, other
}

enum ReportError: LocalizedError {
    case invalidResponse
    case server(statusCode: Int)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Risposta non valida dal server."
        case .server(let statusCode):
            return "Il server ha risposto con codice \(statusCode)."
        }
    }
}

private struct ReportPayload: Encodable {
    let phone_number: String
    let category: String
    let device_hash: String
    let app_platform: String
}

enum ReportService {
    static func submit(phoneNumberE164: String, category: ReportCategory) async throws {
        var request = URLRequest(url: SupabaseConfig.url.appendingPathComponent("rest/v1/reports"))
        request.httpMethod = "POST"
        request.setValue(SupabaseConfig.publishableKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(SupabaseConfig.publishableKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("return=minimal", forHTTPHeaderField: "Prefer")

        let payload = ReportPayload(
            phone_number: phoneNumberE164,
            category: category.rawValue,
            device_hash: DeviceIdentity.hashedDeviceId,
            app_platform: "ios"
        )
        request.httpBody = try JSONEncoder().encode(payload)

        let (_, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ReportError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            throw ReportError.server(statusCode: httpResponse.statusCode)
        }
    }
}
